azconfig/client/src/main/java/com/azure/azconfig/AzConfigCredentialsPolicy.java

```diff
@@ -0,0 +1,142 @@
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Base64;
+
+/**
+ * Creates a policy that authenticates request with AzConfig service.
+ */
+public final class AzConfigCredentialsPolicy implements HttpPipelinePolicy {
+    private static final String SIGNED_HEADERS = "host;x-ms-date;x-ms-content-sha256";
+    private static final String KEY_VALUE_APPLICATION_HEADER = "application/vnd.microsoft.azconfig.kv+json";
+
+    private static final String HOST_HEADER = "Host";
+    private static final String DATE_HEADER = "x-ms-date";
+    private static final String CONTENT_HASH_HEADER = "x-ms-content-sha256";
+    private static final String AUTHORIZATION_HEADER = "Authorization";
+    private static final String CONTENT_TYPE_HEADER = "Content-Type";
+    private static final String ACCEPT_HEADER = "Accept";
+
+    private final AzConfigClient.AzConfigCredentials credentials;
+    private final HttpPipelineOptions options;
+
+    /**
+     * Initializes a new instance of AzConfigCredentialsPolicy based on credentials.
+     *
+     * @param credentials for the Configuration Store in Azure
+     */
+    AzConfigCredentialsPolicy(AzConfigClient.AzConfigCredentials credentials) {
+        this(credentials, new HttpPipelineOptions(null));
+    }
+
+    /**
+     * Initializes a new instance of AzConfigCredentialsPolicy based on credentials.
+     *
+     * @param credentials for the Configuration Store in Azure
+     * @param options the request options
+     */
+    AzConfigCredentialsPolicy(AzConfigClient.AzConfigCredentials credentials, HttpPipelineOptions options) {
+        this.credentials = credentials;
+        this.options = options;
+    }
+
+    /**
+     * Sign the request.
+     *
+     * @param context The request context
+     * @return A {@link Mono} representing the HTTP response that will arrive asynchronously.
+     */
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, NextPolicy next) {
+        context.httpRequest().headers().set(HOST_HEADER, credentials.baseUri().getHost());
+        if (context.httpRequest().headers().value(DATE_HEADER) == null) {
+            OffsetDateTime now = Instant.now().atOffset(ZoneOffset.UTC);
+            String utcNowString = DateTimeFormatter.RFC_1123_DATE_TIME.toFormat().format(now);
+            context.httpRequest().headers().set(DATE_HEADER, utcNowString);
+        }
+        String contentHash = getContentHash(context.httpRequest().body() == null ? new byte[0] : context.httpRequest().body().blockFirst().array());
+        context.httpRequest().headers().set(CONTENT_HASH_HEADER, contentHash);
+        context.httpRequest().headers().set(CONTENT_TYPE_HEADER, KEY_VALUE_APPLICATION_HEADER);
+        context.httpRequest().headers().set(ACCEPT_HEADER, KEY_VALUE_APPLICATION_HEADER);
```

Comment:
I'm not really doing an impl review right now, but this kind of code sticks out to me - we should grab the headers once and not do the whole call chain every time.